wgsl/index.bs

```diff
@@ -3346,29 +3346,94 @@ and vice versa,
 where |p| and |r| describe the same memory view.
 </blockquote>
 
+### Valid and Invalid Memory References ### {#valid-invalid-memory-references}
+
+A [=reference type|reference=] value is either [=valid reference|valid=] or [=invalid memory reference|invalid=].
+
+References are formed as described in detail in [[#forming-references-and-pointers]].
+Generally, a <dfn noexport>valid reference</dfn> is formed by:
+* naming a variable, or
+* applying the [=indirection=] (unary `*`) operation to a [=valid pointer=], or
+* forming a [=composite reference component expression=] around a valid reference,
+    where the index in an indexing phrase, if used, is within bounds for the type.
+
+Generally, an <dfn noexport>invalid memory reference</dfn> is formed by:
+* applying the indirection operator to an [=invalid pointer=], or
+* forming a [=composite reference component expression=] around an invalid reference,
+    or by indexing with an out-of-bounds index.
+
+A pointer value always corresponds to a reference value, so an
+<dfn noexport>valid pointer</dfn> is a pointer that corresponds to a [=valid reference=],
+and an <dfn noexport>invalid pointer</dfn> is a pointer that corresponds to an [=invalid memory reference=].
+
 ### Originating Variable ### {#originating-variable-section}
 
-In WGSL a reference value always corresponds to the memory view
-for some or all of the memory locations for some variable.
+A [=valid reference=] corresponds to the memory view for some or all of the memory locations for some variable.
 This defines the <dfn noexport>originating variable</dfn> for the reference value.
 
-A pointer value always corresponds to a reference value, and so the originating variable
-of a pointer is the same as the originating variable of the corresponding reference.
+The [=originating variable=] of a [=valid pointer=] is the same as the originating variable of the corresponding valid reference.
```

Comment:
Is the comment about *pointer* or do you mean it to apply to *reference* as well.

I was bothered with the idea that an invalid reference could be assumed to be associated with memory locations. It shouldn't, because the shader shouldn't access those locations.

 I guess you could say you're only allowed to access the locations if the reference is valid.  But that's keeping an extra bool around.  Also, the spec shouldn't prescribe *which* memory locations should be associated with an invalid reference.  I want to keep it simple by saying an invalid reference (or pointer, respectively), are not associated with locations.

And the same logic applies to only valid references (pointers) having an originating variale.

Reply:
If we can't find the originating variable then we have to change the logic about what pointer parameters are allowed for example. To me, the originating variable is more about variable provenance than locations.

Reply:
got it.  I managed to rewrite this and collapse a bunch of text elsewhere.